mautic/mautic
# Segment refactoring

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/262
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

We added new code to handle segments. Old methods are still available, but deprecated.
This PR is focused on code refactoring and to replace subSelect by JOINs in SQL. New queries should be faster then old ones in most cases.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. Make sure you have database with lot of data (segments and leads).
2. Update all your segments from console
3. Checkout this branch and clear cache.
4. Run command `app/console mautic:segments:check-builders --env=dev`. It compares old result to new ones. You can specify segment's id by `--segment-id=1` parameter if you hit limits or need to explore any segment.
5. We have fixed a few bugs in old code (e.g. handling relative dates), so results could be little different. You need to compare old SQL to new one and find if it a bug. You can find both queries in log.
6. We have several different query builders. We need least 1 test for each of these filters:
- Email read count
- Email received
- Email read date
- Hit URL count
- Hit URL date
- DNC bounced
- DNC unsubscribed
- Device type / brand
- UTM tags
7. Run command for segments update. No error should occur and No changes in segment should be found.
8. Clear all `lead_lists_leads` data and run segment update again. All data should be processed correctly.

#### List deprecations along with the new alternative:
1.  `LeadListRepository / getLeadsByList()` > `ContactSegmentService` and proper method (`getNewLeadListLeadsCount`, `getTotalLeadListLeadsCount`, `getNewLeadListLeads`, `getOrphanedLeadListLeadsCount`, `getOrphanedLeadListLeads`
2. `ListModel / getLeadsByList` > `ContactSegmentService` and proper method

#### List backwards compatibility breaks:
1. 
2.
